### ISSUE TITLE
Implement scale endpoint for jobs

### ISSUE THREAD

**Issue by soltysh** [kubernetes/kubernetes#38756]
Currently Deployments, RC and RS have scale endpoint which allows directly scaling those instead of mutating the object's spec when scaling up/down. `kubectl scale job` modifies `.spec.parallelism` directly. I'm proposing jobs should have `/scale` endpoint to leverage scale functionality. 

@kubernetes/sig-apps ideas/objections?

**Comment by kargakis** [comment]
SGTM

**Comment by AdoHe** [comment]
SGTM.

**Comment by wanghaoran1988** [comment]
@soltysh @kargakis I am try to add the scale endpoint for the Job,  but I have a problem now, I am using the scale struct in the [api extentions pkg](https://github.com/wanghaoran1988/kubernetes/blob/8c1f45953b7c678f16ccd25bc0c6c8e4a5ea53d1/staging/src/k8s.io/client-go/pkg/apis/extensions/types.go#L68)，I am not sure whether it's correct, or should I create a new one in batch pkg ?
Also I found the `parallelism` filed doesn't exits in the `JobStatus` struct, should we update to add this to the struct to show how many replicas are actually running?

**Comment by kargakis** [comment]
> Also I found the parallelism filed doesn't exits in the JobStatus struct, should we update to add this to the struct to show how many replicas are actually running?

Yeah, I have thought about this before and I think it deserves its own issue. Can you please open one and discuss that there?

**Comment by wanghaoran1988** [comment]
After I go through the logic, I find that actual parallelism (number of pods running at any instant) may be more or less than requested parallelism,[docs](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#controlling-parallelism). So I think to align with the doc, we need have it's own Scale struct, here is the spec in my mind, the `Completions` field is immutable, but the real parallelism pods depends on this, so I put that in the `ScaleStatus` spec.
@kargakis @soltysh Could you please take a look, if it's ok, I will start working on this.
```
type ScaleSpec struct {
	// Specifies the maximum desired number of pods the job should
	// run at any given time. The actual number of pods running in steady state will
	// be less than this number when ((.spec.completions - .status.successful) < .spec.parallelism),
	// i.e. when the work left to do is less than max parallelism.
	// +optional
	Parallelism *int32
}

// represents the current status of a scale subresource.
type ScaleStatus struct {
	// Specifies the desired number of successfully finished pods the
	// job should be run with.  Setting to nil means that the success of any
	// pod signals the success of all pods, and allows parallelism to have any positive
	// value.  Setting to 1 means that parallelism is limited to 1 and the success of that
	// pod signals the success of the job.
	// +optional
	Completions *int32

	// The number of actively running pods.
	// +optional
	Active int32

	// label query over pods that should match the replicas count.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
	// +optional
	Selector *metav1.LabelSelector
}

// represents a scaling request for a resource.
type Scale struct {
	metav1.TypeMeta
	// Standard object metadata; More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata.
	// +optional
	metav1.ObjectMeta

	// defines the behavior of the scale. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
	// +optional
	Spec ScaleSpec

	// current status of the scale. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status. Read-only.
	// +optional
	Status ScaleStatus
}
```

**Comment by kargakis** [comment]
@soltysh can you ptal on @wanghaoran1988's proposal?

**Comment by soltysh** [comment]
@wanghaoran1988 here's the thing. I've talked some time ago with @deads2k and we both agreed that having `Scale` structure inside every API group isn't the right solution. The idea was to have one `Scale` structure that would work with every resource out there. To do that we'd need some dynamic scale client, see https://github.com/kubernetes/kubernetes/pull/32523#issuecomment-267050916. If you could work on it I'd be more than happy reviewing it :)

**Comment by deads2k** [comment]
I have a pull adding the required information to discovery here: https://github.com/kubernetes/kubernetes/pull/49971

**Comment by soltysh** [comment]
There's also https://github.com/kubernetes/kubernetes/issues/29698 touching this topic.

**Comment by wanghaoran1988** [comment]
Seems @DirectXMan12 is working on this from the comments here:https://github.com/kubernetes/kubernetes/pull/49971#issuecomment-319496553

**Comment by wanghaoran1988** [comment]
@DirectXMan12 Are you working on this?

**Comment by DirectXMan12** [comment]
Once #49971 lands, I'll revive my polymorphic scale client PR in some form.

**Comment by sttts** [comment]
@DirectXMan12 https://github.com/kubernetes/kubernetes/pull/49971 landed. Any plans for the polymorphic scale?

**Comment by DirectXMan12** [comment]
Polymorphic scale landed in 1.9, FWIW (see the referenced commits in the activity stream above), so this is no longer bloked by that

**Comment by soltysh** [comment]
Based on https://github.com/kubernetes/kubernetes/pull/58468#issuecomment-360794229 this is not going to happen. Additionally, we're starting to deprecate `kubectl scale job` functionality in 1.10.

**Comment by cometta** [comment]
Will there be a new scale endpoint feature after deprecated scale job command?

**Comment by soltysh** [comment]
Nope, scaling job was considered a mistake, since they are different from all other resources that allow scaling. You can still modify it's spec field, though.

**Comment by shangmu** [comment]
Then what's the expected behavior (or fundamental difference) of modifying the spec field? If there's no fundamental difference in the two (between `kubectl scale` and editing the spec field), why not just keep the `scale job` command and explicitly state the limitations?
